homeassistant/components/binary_sensor/xboxlive.py

```diff
@@ -0,0 +1,172 @@
+        self._xuid = gamer_profile.xuid
+        self._picture = gamer_profile.gamerpic
+        self._headers = {
+            'x-xbl-contract-version': '2',
+            'User-Agent': ('XboxRecord.Us Like SmartGlass/2.105.0415 '
+                           'CFNetwork/711.3.18 Darwin/14.0.0')}
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return "xboxlive_{}".format(self._gamertag)
+
+    @property
+    def is_on(self):
+        """Return the status of the sensor."""
+        return self._state == 'online'
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        attributes = {
+            "xuid": self._xuid,
+            "gamerscore": self._gamerscore,
+        }
+        for device in self._devices:
+            for title in device.get('titles'):
+                attributes[
+                    '{}_{}'.format(device.get('type').lower(),
+                                   title.get('placement').lower())
+                ] = title.get('name')
+
+        return attributes
+
+    @property
+    def entity_picture(self):
+        """Avatar of the account."""
+        return self._picture
+
+    @property
+    def icon(self):
+        """Return the icon to use in the frontend."""
+        return ICON
+
+    @asyncio.coroutine
+    def async_update(self):
+        """Update state data from Xbox API."""
+        import xbox
+        # TODO check of client is connected
+        # Get profile
+        gamer_profile = xbox.GamerProfile.from_xuid(self._xuid)
+        self._gamer_profile = gamer_profile
+        self._gamertag = gamer_profile.gamertag
+        self._picture = gamer_profile.gamerpic
+
+        # Get presence
+        url = "https://userpresence.xboxlive.com/users/batch"
+        data = {"users": [self._xuid], "level": "all"}
+        response = xbox.client._post_json(url, data=data,
+                                          headers=self._headers)
+        if response.status_code != 200:
+            _LOGGER.error("Can not get %s presence", self._gamertag)
+            return
+        results = response.json()
+        if len(results) != 1:
+            _LOGGER.error("Can not get %s presence", self._gamertag)
+            return
+
+        # Save state
+        self._state = results[0].get('state', STATE_UNKNOWN).lower()
+        # Save devices
+        self._devices = results[0].get('devices', [])
+
+        # Get Other data
+        url = "https://profile.xboxlive.com/users/batch/profile/settings"
+        settings = ["Gamertag", "RealName", "Bio", "Location", "Gamerscore",
+                    "GameDisplayPicRaw", "AccountTier", "XboxOneRep",
+                    "PreferredColor"]
+        data = {"userIds": [self._xuid], "settings": settings}
+        response = xbox.client._post_json(url, data=data,
```

Comment:
You could extract this including the error checking into a helper method, since you do almost the same thing below. The method should accept the url and data, and return the response. Then you do the custom parsing of the response here outside of the helper method.

Reply:
It's also weird that use access a private method on the client. Is there no other way of getting this info? Why are those methods marked as private in the xbox library?